### ISSUE TITLE
DL4US

### ISSUE THREAD

**Issue by kmotohas** [kmotohas/dl4me#5]
https://github.com/ailorg/dl4us_pub
https://app.ilect.net/

**Comment by kmotohas** [comment]
# Lesson1 手書き文字認識をしよう（ニューラルネットワーク入門）

## 目次

- Section1 解説
  - 1.1 Keras実装プロセス
  - 1.2 各モデルLayer
  - 1.3 損失関数
  - 1.4 評価関数
  - 1.5 Functional API
  - 1.6 確認問題
- Section2 実装①
  - 2.1 MNISTによるMLPの復習
- Section3 テクニック・発展内容
  - 3.1 前処理
  - 3.2 勾配に関するテクニック
    - 3.2.1 最適化アルゴリズム (optimizer)
    - 3.2.2 活性化関数 (activation)
    - 3.2.3 初期化 (initializer)
  - 3.3 過学習に関するテクニック
    - 3.3.1 正則化 (regularization)
    - 3.3.2 早期終了 (early stopping)
    - 3.3.3 ドロップアウト (dropout)
  - 3.4 確認問題
- Section4 実装②
  - 4.1 Fashion MNIST
  - 4.2 実装

**Comment by kmotohas** [comment]
```def homework(train_X, train_y, test_X):
    import keras
    from keras.models import Sequential, Model
    from keras.layers import Dense, Conv2D, MaxPooling2D, Flatten, Input, Activation, Add
    from keras.callbacks import EarlyStopping
    from keras.layers.normalization import BatchNormalization

    inputs = Input(shape=(32, 32, 3))

    model = Sequential()
    model.add(Conv2D(16, kernel_size=(3, 3), padding='same', activation='relu', kernel_initializer='he_normal', input_shape=(32, 32, 3)))
    model.add(Conv2D(16, kernel_size=(3, 3), padding='same', activation='relu', kernel_initializer='he_normal'))
    model.add(MaxPooling2D(pool_size=(2, 2)))
    model.add(Conv2D(32, kernel_size=(3, 3), padding='same', activation='relu', kernel_initializer='he_normal'))
    model.add(Conv2D(32, kernel_size=(3, 3), padding='same', activation='relu', kernel_initializer='he_normal'))
    model.add(MaxPooling2D(pool_size=(2, 2)))
    model.add(Conv2D(64, kernel_size=(3, 3), padding='same', activation='relu', kernel_initializer='he_normal'))
    model.add(Conv2D(64, kernel_size=(3, 3), padding='same', activation='relu', kernel_initializer='he_normal'))
    model.add(Conv2D(64, kernel_size=(3, 3), padding='same', activation='relu', kernel_initializer='he_normal'))
    model.add(Conv2D(64, kernel_size=(3, 3), padding='same', activation='relu', kernel_initializer='he_normal'))
    model.add(MaxPooling2D(pool_size=(2, 2)))
    model.add(Flatten())
    model.add(Dense(128, activation='sigmoid'))
    model.add(Dense(10, activation='softmax'))
    
    model.compile(loss=keras.losses.categorical_crossentropy, optimizer='adam', metrics=['accuracy'])
    
    early_stopping = EarlyStopping(patience=1, verbose=1)
    
    model.fit(x=train_X, y=train_y, batch_size=256, epochs=32, validation_split=0.1, callbacks=[early_stopping])
    
    pred_y = model.predict(test_X)
    pred_y = np.argmax(pred_y, 1)
    
    return pred_y
```
```
Train on 45000 samples, validate on 5000 samples
Epoch 1/32
45000/45000 [==============================] - 18s 401us/step - loss: 1.7429 - acc: 0.3558 - val_loss: 1.4515 - val_acc: 0.4694
Epoch 2/32
45000/45000 [==============================] - 15s 342us/step - loss: 1.3124 - acc: 0.5256 - val_loss: 1.3326 - val_acc: 0.5306
Epoch 3/32
45000/45000 [==============================] - 15s 343us/step - loss: 1.1381 - acc: 0.5971 - val_loss: 1.0999 - val_acc: 0.6062
Epoch 4/32
45000/45000 [==============================] - 15s 341us/step - loss: 1.0061 - acc: 0.6455 - val_loss: 1.0257 - val_acc: 0.6360
Epoch 5/32
45000/45000 [==============================] - 15s 342us/step - loss: 0.8928 - acc: 0.6858 - val_loss: 0.9633 - val_acc: 0.6624
Epoch 6/32
45000/45000 [==============================] - 15s 342us/step - loss: 0.8082 - acc: 0.7186 - val_loss: 0.8922 - val_acc: 0.6850
Epoch 7/32
45000/45000 [==============================] - 15s 342us/step - loss: 0.7471 - acc: 0.7396 - val_loss: 0.8730 - val_acc: 0.6944
Epoch 8/32
45000/45000 [==============================] - 15s 342us/step - loss: 0.6865 - acc: 0.7606 - val_loss: 0.9097 - val_acc: 0.6876
Epoch 00008: early stopping
0.694612322577
```
